action/action-simulator/src/main/java/com/powsybl/action/simulator/loadflow/LoadFlowActionSimulator.java

```diff
@@ -246,4 +255,88 @@ public boolean isActionTaken(String actionId) {
             return false;
         }
     }
+
+    private void runTests(ActionDb actionDb, RunningContext context) {
+        // test actions
+        EvaluationContext evalContext = new EvaluationContext() {
+            @Override
+            public Network getNetwork() {
+                return context.getNetwork();
+            }
+
+            @Override
+            public Contingency getContingency() {
+                return context.getContingency();
+            }
+
+            @Override
+            public boolean isActionTaken(String actionId) {
+                return context.getTimeLine().actionTaken(actionId);
+            }
+        };
+
+        List<Rule> activedRules = actionDb.getRules().stream()
+                .filter(rule -> rule.getType().equals(RuleType.TEST))
+                .filter(rule -> {
+                    ExpressionNode conditionExpr = ((ExpressionCondition) rule.getCondition()).getNode();
+                    return ExpressionEvaluator.evaluate(conditionExpr, evalContext).equals(Boolean.TRUE);
+                })
+                .collect(Collectors.toList());
+        List<String> testActionIds = activedRules.stream()
+                                .flatMap(r -> r.getActions().stream())
+                                .distinct()
+                                .filter(id -> !context.isTested(id))
+                                .collect(Collectors.toList());
+        byte[] contextNetwork = NetworkXml.gzip(context.getNetwork());
+
+        for (String actionId : testActionIds) {
+            Action action = actionDb.getAction(actionId);
+            Network networkForTest = NetworkXml.gunzip(contextNetwork);
+            LoadFlowResult testResult = runTest(context, networkForTest, action);
+            context.addTested(actionId);
+            if (testResult.isOk()) {
+                List<LimitViolation> violationsInTest =
+                        LIMIT_VIOLATION_FILTER.apply(Security.checkLimits(networkForTest, 1), networkForTest);
+                if (violationsInTest.isEmpty()) {
+                    context.addWorkedTest(action.getId());
+                    if (applyIfWorks) {
+                        LOGGER.info("Loadflow with test {} works already and exists", action.getId());
+                        observers.forEach(o -> o.noMoreViolationsAfterTest(context, action.getId()));
+                        observers.forEach(o -> o.beforeApplyTest(context, action.getId()));
+                        action.run(context.getNetwork(), computationManager);
+                        context.getTimeLine().getActions().add(actionId);
+                        observers.forEach(o -> o.loadFlowConverged(context, violationsInTest));
+                        observers.forEach(o -> o.noMoreViolations(context));
+                        observers.forEach(o -> o.afterApplyTest(context, action.getId()));
```

Comment:
It seems that some events are sent lately: I think that the beforeApplyTest event should be notified earlier, don’t you?